sparse/_coo/common.py

```diff
@@ -1287,6 +1299,42 @@ def _dot_ndarray_coo(array1, coords2, data2, out_shape):  # pragma: no cover
     return _dot_ndarray_coo
 
 
+@_memoize_dtype
+def _dot_ndarray_coo_type_sparse(dt1, dt2):
+    dtr = np.result_type(dt1, dt2)
+
+    @numba.jit(nopython=True, nogil=True)
+    def _dot_ndarray_coo(array1, coords2, data2, out_shape):  # pragma: no cover
+        """
+        Utility function taking in two one ``ndarray`` and one ``COO`` and
+        calculating a "sense" of their dot product. Acually computes ``x1 @ s2.T``.
+
+        Parameters
+        ----------
+        array1 : np.ndarray
+            The input array ``x1``.
+
+        data2, coords2 : np.ndarray
+            The data and coordinates of ``s2``.
+
+        out_shape : Tuple[int]
+            The output shape.
+        """
+        out_data = []
+        out_coords = []
+
+        for oidx1 in range(out_shape[0]):
```

Comment:
I'd perform the accumulation inline rather than relying on the deduplication/coalescing. I believe this is what is slowing down your performance.

Reply:
Thanks for the suggestions! Could you please elaborate on this one?

Reply:
Well, matrix multiplies are a multiply of some elements plus a sum over a given axis, right? What you're doing is adding all the multiplies in, and relying on the deduplication to do the sum. I would also do the sum here as well, and filter out zero entries from the get-go. This would improve performance by a lot.

Reply:
I am not sure what you mean exactly, but the summing seems to be optimal already because for every row of the output (oidx1) we only iterate over the nonzero elements of the sparse array once (didx). How can you improve that? Could you provide some code or formula for your idea?

To the filtering out: I already tried that and it didnt seem to have any effect on the performance in the case that I tested. As the first array is dense, we should assume that it does not have a lot of zero entries, maybe none at all. Thus, the test for zero would almost always be negative. The additional overhead created by this test seems to be of the same order as the one call of prune in the creation of the COO array.

If you prefer, I can switch to the version with tests for zero and without prune, but I have not seen a performance increase in that change.

Reply:
You are adding way more elements than necessary IIUC, one element is added multiple times. If you follow my suggestion in the other comment you'll see that.

Reply:
You are right! I updated the algorithm accordingly and the performance now matches the return_type=None version.
Unfortunately the test fails in Python 3.6. Any idea why?